Docs builder extensions #162 (#163)
ardalis/Specification: docs/extensions/create-specification-builder.md

nav_order: 2
---

# How to create your own specification builder

> How to create your own specification builder 

Reviewer: ```suggestion

How to create your own specification builder 
```

Reviewer: Also, although the title is all about "How" it would be great to say "Why" one would want/need to do this somewhere at the top, too.

## Example: Configure caching behaviour through specification builder extension method

> In order to achieve this:

Reviewer: ```suggestion
In order to achieve this (note the `.WithTimeToLive` method):
```

````csharp
public class CustomerByNameWithStores : Specification<Customer>
{
    public CustomerByNameWithStores(string name)
    {
        Query.Where(x => x.Name == name)
            .EnableCache(nameof(CustomerByNameWithStoresSpec), name)
                // Can only be called after .EnableCache()
                .WithTimeToLive(TimeSpan.FromHours(1))
            .Include(x => x.Stores);
    }
}
````

We can create a simple extension method on the specification builder:

````csharp
public static class SpecificationBuilderExtensions
{
    public static ISpecificationBuilder<T> WithTimeToLive<T>(this ISpecificationBuilder<T> @this, TimeSpan ttl)
        where T : class
    {
        @this.Specification.SetCacheTTL(ttl);
        return @this;
    }
}
````

This extension method can only be called when chained after `SpecificationBuilderExtensions.EnableCache`. This is because `EnableCache` returns `ICacheSpecificationBuilder<T>` which inherits from `ISpecificationBuilder<T>`.

```csharp
// TODO: Repository example
```

> Finally, we need to take of some plumbing to implement both `` and ``. The class below uses `ConditionalWeakTable` to do the trick. An other solution is to create a base class that inherits from `Specification<T>`.

Reviewer: ```suggestion
Finally, we need to take of some plumbing to implement both `` and ``. The class below uses `ConditionalWeakTable` to do the trick. Another solution is to create a base class that inherits from `Specification<T>`.
```

````csharp

> public static class SpecificationExtentions

Reviewer: ```suggestion
public static class SpecificationExtensions
```
